pkg/gadgets/socket-collector/gadget.go

```diff
@@ -0,0 +1,140 @@
+// Copyright 2021 The Inspektor Gadget authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package socketcollector
+
+import (
+	"fmt"
+	"sync"
+
+	log "github.com/sirupsen/logrus"
+	"k8s.io/apimachinery/pkg/types"
+
+	gadgetv1alpha1 "github.com/kinvolk/inspektor-gadget/pkg/api/v1alpha1"
+	"github.com/kinvolk/inspektor-gadget/pkg/gadgets"
+	"github.com/kinvolk/inspektor-gadget/pkg/gadgets/socket-collector/tracer"
+)
+
+type Trace struct {
+	resolver gadgets.Resolver
+}
+
+type TraceFactory struct {
+	gadgets.BaseFactory
+	mu     sync.Mutex
+	traces map[string]*Trace
+}
+
+func setStatusError(trace *gadgetv1alpha1.Trace, err string) {
+	trace.Status.OperationError = err
+	trace.Status.Output = ""
+	trace.Status.State = "Stopped"
+}
+
+func (f *TraceFactory) LookupOrCreate(name types.NamespacedName) gadgets.Trace {
+	f.mu.Lock()
+	defer f.mu.Unlock()
+	if f.traces == nil {
+		f.traces = make(map[string]*Trace)
+	}
+	trace, ok := f.traces[name.String()]
+	if ok {
+		return trace
+	}
+	trace = &Trace{
+		resolver: f.Resolver,
+	}
+	f.traces[name.String()] = trace
+
+	return trace
+}
+
+func (f *TraceFactory) Delete(name types.NamespacedName) error {
+	log.Infof("Deleting %s", name.String())
+	f.mu.Lock()
+	defer f.mu.Unlock()
+	_, ok := f.traces[name.String()]
+	if !ok {
+		log.Infof("Deleting %s: does not exist", name.String())
+		return nil
+	}
+	delete(f.traces, name.String())
+	return nil
+}
+
+func (t *Trace) Operation(trace *gadgetv1alpha1.Trace,
+	operation string,
+	params map[string]string) {
+
+	if trace.ObjectMeta.Namespace != gadgets.TRACE_DEFAULT_NAMESPACE {
+		setStatusError(trace, fmt.Sprintf("This gadget only accepts operations on traces in the %s namespace", gadgets.TRACE_DEFAULT_NAMESPACE))
+		return
+	}
+
+	switch operation {
+	case "start":
+		var pid uint32 = 0
+
+		if trace.Spec.Filter.Namespace == "" {
+			setStatusError(trace, "Invalid filter: missing namespace")
+			return
+		}
+
+		if trace.Spec.Filter.Podname == "" {
+			setStatusError(trace, "TODO: Filtering only by namespace is not currently supported")
+			return
+		}
+
+		if trace.Spec.Filter.ContainerName != "" {
+			log.Printf("Gadget %s: Container name filter is not applicable in this gadget, ignoring it!", trace.Spec.Gadget)
+		}
```

Comment:
Should you add a similar check for `trace.Spec.Filter.Labels`? E.g. seccomp has:
```
        if trace.Spec.Filter.Labels != nil {
                trace.Status.OperationError = "Seccomp gadget does not support filtering by labels"
                return
        }
```

Reply:
Done. Added also in TODO list for the future.